### PR TITLE
kvencoder: fix allocator rebase to a smaller newBase

### PR DESCRIPTION
**Pull request by winkyao**
@huachaohuang @holys @coocood PTAL

### CODE REVIEW

File: util/kvencoder/allocator.go [hunk 1]
```diff
@@ -19,38 +19,55 @@ import (
...
+		if newBase <= oldBase {
+			break
+		}
+		if atomic.CompareAndSwapInt64(&alloc.base, oldBase, newBase) {
+			break
+		}
+	}
+
 	return nil
 }
```
Comment (holys) [inline util/kvencoder/allocator.go]: Add comment for it to fix golint

File: util/kvencoder/allocator.go [hunk 2]
```diff
@@ -19,38 +19,55 @@ import (
...
 
-func (alloc *allocator) Base() int64 {
+func (alloc *Allocator) Base() int64 {
```
Comment (holys) [inline util/kvencoder/allocator.go]: Add comment for it to fix golint

File: util/kvencoder/allocator.go [hunk 3]
```diff
@@ -19,38 +19,55 @@ import (
...
 	return atomic.LoadInt64(&alloc.base)
 }
 
-func (alloc *allocator) End() int64 {
+func (alloc *Allocator) End() int64 {
```
Comment (holys) [inline util/kvencoder/allocator.go]: ditto

File: util/kvencoder/allocator.go [hunk 4]
```diff
@@ -19,38 +19,55 @@ import (
...
 	return alloc.Base() + step
 }
 
-func (alloc *allocator) NextGlobalAutoID(tableID int64) (int64, error) {
+func (alloc *Allocator) NextGlobalAutoID(tableID int64) (int64, error) {
```
Comment (holys) [inline util/kvencoder/allocator.go]: ditto

**Comment by coocood** [comment]
@winkyao 
The test is under the same package name, we don't need to export the allocator.

**Comment by winkyao** [comment]
@coocood Allocator.ResetBase maybe used by lightning.

File: util/kvencoder/allocator.go [hunk 5]
```diff
@@ -19,38 +19,59 @@ import (
...
 
-func (alloc *allocator) Alloc(tableID int64) (int64, error) {
+// Alloc allocs the next autoID for table with tableID.
+func (alloc *Allocator) Alloc(tableID int64) (int64, error) {
 	return atomic.AddInt64(&alloc.base, 1), nil
 }
 
-func (alloc *allocator) Rebase(tableID, newBase int64, allocIDs bool) error {
+// ResetBase allow newBase smaller than alloc.base, and will set the alloc.base to newBase.
+func (alloc *Allocator) ResetBase(newBase int64) {
```
Comment (zhexuany) [inline util/kvencoder/allocator.go]: why do you want to change the function name?

Comment (winkyao) [reply util/kvencoder/allocator.go]: the func is added, not changed.

**Comment by holys** [comment]
LGTM

File: util/kvencoder/allocator.go [hunk 6]
```diff
@@ -19,38 +19,59 @@ import (
...
-// NewAllocator new an allocator.
-func NewAllocator() autoid.Allocator {
-	return &allocator{}
+// NewAllocator new an Allocator.
+func NewAllocator() *Allocator {
+	return &Allocator{}
 }
 
-type allocator struct {
+// Allocator is a id allocator, it is only used in lightning.
```
Comment (huachaohuang) [inline util/kvencoder/allocator.go]: s/a/an

File: util/kvencoder/allocator.go [hunk 7]
```diff
@@ -19,38 +19,59 @@ import (
...
+var _ autoid.Allocator = &Allocator{}
 
 var (
 	step = int64(5000)
 )
 
-// NewAllocator new an allocator.
-func NewAllocator() autoid.Allocator {
-	return &allocator{}
+// NewAllocator new an Allocator.
```
Comment (huachaohuang) [inline util/kvencoder/allocator.go]: s/new/creates

File: util/kvencoder/allocator.go [hunk 8]
```diff
@@ -19,38 +19,59 @@ import (
...
 }
 
-type allocator struct {
+// Allocator is a id allocator, it is only used in lightning.
+type Allocator struct {
 	base int64
 }
 
-func (alloc *allocator) Alloc(tableID int64) (int64, error) {
+// Alloc allocs the next autoID for table with tableID.
```
Comment (huachaohuang) [inline util/kvencoder/allocator.go]: s/the/a

File: util/kvencoder/allocator.go (hunk 5 above)
Comment (huachaohuang) [reply util/kvencoder/allocator.go]: I think "Reset" is clear enough.

File: util/kvencoder/allocator.go [hunk 9]
```diff
@@ -19,38 +19,59 @@ import (
...
+		if atomic.CompareAndSwapInt64(&alloc.base, oldBase, newBase) {
+			break
+		}
+	}
+
 	return nil
 }
 
-func (alloc *allocator) Base() int64 {
+// Base return the current base of Allocator.
```
Comment (huachaohuang) [inline util/kvencoder/allocator.go]: s/return/returns

**Comment by huachaohuang** [comment]
@holys We should use `Reset` instead of `Rebase` in lightning.

**Pull request merged**